apm-agent-core/src/main/java/co/elastic/apm/agent/util/JmxUtils.java

```diff
@@ -0,0 +1,95 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.util;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.lang.management.OperatingSystemMXBean;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A utility to obtain JVM-specific MBean implementations, supporting HotSpot and J9.
+ *
+ * This implementation is based on io.micrometer.core.instrument.binder.system.ProcessorMetrics,
+ * under Apache License 2.0
+ */
+public class JmxUtils {
+
+    private static final Logger logger = LoggerFactory.getLogger(JmxUtils.class);
+
+    private static boolean initialized;
+
+    /**
+     * List of public, exported interface class names from supported JVM implementations.
+     */
+    private static final List<String> OPERATING_SYSTEM_BEAN_CLASS_NAMES = Arrays.asList(
+        // NOTE: THE ORDER IS IMPORTANT AS J9 CONTAINS THE SUN INTERFACE AS WELL
+        "com.ibm.lang.management.OperatingSystemMXBean", // J9
+        "com.sun.management.OperatingSystemMXBean" // HotSpot
+    );
+
+    @Nullable
+    private static Class<?> operatingSystemBeanClass;
+
+    @Nullable
+    public synchronized static Method getOperatingSystemMBeanMethod(OperatingSystemMXBean operatingSystemBean, String methodName) {
+        if (!initialized) {
+            // lazy initialization - try loading the classes as late as possible
```

Comment:
why do we need to do that ? can you elaborate a bit when this is required and how it helps ?
For example, do we need to wait another part of the application (like an application server startup) to execute before trying to use this code ?

Reply:
> why do we need to do that ?

Because it is always a good idea to look for a class in the classpath as late as possible, as I elaborated. Otherwise this check may be premature.

> For example, do we need to wait another part of the application (like an application server startup) to execute before trying to use this code ?

Maybe 🙂 . I rather not wait for someone reporting on such scenario. Unless there is a good reason to avoid, safer code seems like the better option.